### ISSUE TITLE
Optional support for unsecure keystone API URL

### ISSUE THREAD

**Issue by oxygen0211** [kubernetes/kubernetes#22695]
Hi,
I tried to integrate the keystone auth plugin (which would be perfect for my needs) into the cluster I'm currently setting up. However, I can't get the connection working since our keystone installation is currently only reachable over http

> server.go:492] Invalid Authentication Config: Auth URL should be secure and start with https

As far as I can see in plugin/pkg/auth/authenticator/request/keystone/keystone.go, there is no option to deactivate this check. I think this should be an optional parameter on startup, similar to `--experimental-keystone-url` defaulting to the check being activated.

**Comment by liggitt** [comment]
I don't think we should enable username/password checking over http.

**Comment by oxygen0211** [comment]
I agree that we are loosing a security layer here, but you can't guarantee you got the password from client over https, so it might be intercepted at this point anyways.

More general, I think the proper way to authenticate to Kubernetes with the Keystone plugin using tokens since they will 

1. Have limited lifetime
2. Be restricted on keystone only which is a big deal regarding keystone likely being bound to a LDAP and thus the password being valid for other services.

So I don't want to check username/password directly. I'm just starting out with the Keystone plugin, so I want to point this out in general, if token auth is supported, the better. Couldn't try it in praxis for now since I can't get the auth URL running.

**Comment by quinton-hoole** [comment]
Assigning to @etune to add a priority label.

**Comment by erictune** [comment]
Setting the priority low until we have a sense of how many users have this restriction (keystone only reachable over http).

**Comment by kfox1111** [comment]
I ran into this same issue with trying to deploy kubernetes with devstack in the same vm and wanted to test integration and do some development. I think there are a significant number of developers in this same boat. http is secure enough since it doesn't leave the developers vm and it isn't intended for production.

The use case is the same as:
https://github.com/kubernetes/kubernetes/issues/24984

**Comment by liggitt** [comment]
I'd vastly prefer allowing specifying the CA for the keystone server over allowing http auth calls carrying usernames and passwords

**Comment by kfox1111** [comment]
I don't disagree. :)

But, I also think Kubernetes shouldn't handle the usernames/passwords at all. In http mode, sending a token is much less risky then a username/password since they expire relatively quickly.

See this issue for details:
https://github.com/kubernetes/kubernetes/issues/24982

**Comment by liggitt** [comment]
I agree tokens would be preferable, but I wouldn't expect credentials of any kind sent over http to remote servers

**Comment by kfox1111** [comment]
Agreed. No one in their right mind should be passing creds or bearer tokens over http over a network. I'm just interested in being able to easily deploy a k8s in DevStack with a simple plugin, where that k8s is hooked into the deployed keystone so that we can start developing things in OpenStack that depend on k8s. This gives developers a smooth/easy experience to then focus on doing the development stuff and not have to mess with setting up a CA/k8s, which will prevent some of them from contributing to k8s/OpenStack. The idea is simple for developers, secure for operators. Totally separate use cases.

**Comment by liggitt** [comment]
> This gives developers a smooth/easy experience to then focus on doing the development stuff and not have to mess with setting up a CA/k8s

It should be as simple as telling k8s to trust the keystone cert as a CA (assuming the self-signed cert generated by keystone is correctly marked as a CA, which it should be, since it is signing for itself).

> The idea is simple for developers, secure for operators. Totally separate use cases.

I've seen too many production installations running with http or untrusted TLS settings to believe that only developers will use this option if it's made available.

**Comment by kfox1111** [comment]
I didn't think Devstack could easily deploy with SSL enabled, but it looks like you can:
https://ask.openstack.org/en/question/45348/devstack-with-ssl/

So http support might not be as important as I thought. It could be relegated to a documentation issue for deploying Devstack in this mode when deploying with Kubernetes.

We'd still need a way to specify the root CA though for Keystone.

**Comment by kfox1111** [comment]
So, I played around with this a little. For my current non devstack based system, I've deployed kubernetes as a cluster running on my laptop deployed as a set of pods via /etc/kubernetes/manifests. Only kubelet is native and all the rest of the services are containers direct from google. Its working well, but since I have 3 api server containers, and get the containers direct from google, I'd have to jump through a bunch of hoops to add the keystone ssl ca cert to each of the containers. I still think an option to specify the ssl cert for keystone would be useful rather then having to merge it into the system's chain of trust.
